web/server/codechecker_server/api/mass_store_run.py

```diff
@@ -61,10 +65,44 @@ def __enter__(self, *args):
         LOG.info("[%s] %s...", self.__run_name, self.__msg)
 
     def __exit__(self, *args):
-        LOG.info("[%s] %s done... (duration: %s sec)", self.__run_name,
+        LOG.info("[%s] %s. Done. (Duration: %s sec)", self.__run_name,
                  self.__msg, round(time.time() - self.__start_time, 2))
 
 
+class _RunLock:
+    def __init__(self, session: DBSession, run_name: str):
+        self.__session = session
+        self.__run_name = run_name
+
+    def __enter__(self, *args):
+        # Load the lock record for "FOR UPDATE" so that the transaction that
+        # handles the run's store operations has a lock on the database row
+        # itself.
+        self.__run_lock = self.__session.query(RunLock) \
+            .filter(RunLock.name == self.__run_name) \
+            .with_for_update(nowait=True) \
+            .one()
+
+        # Do *NOT* remove this seemingly dummy print, we need to make sure
+        # that the execution of the SQL statement is not optimised away and
+        # the fetched row is not garbage collected.
+        LOG.debug("Acquired exclusive lock for run '%s' that was originally "
+                  "locked at '%s'.",
+                  self.__run_name, self.__run_lock.locked_at)
+        return self
+
+    def __exit__(self, *args):
+        self.__run_lock = None
+        self.__session = None
+
+    def noop(self):
+        """
+        Does nothing, but prevents Pylint from complaining, and prevents the
+        context from being optimised out.
```

Comment:
Unused variable, unused parameter... Note that I moved the code over to the context manager here. It would be weird if the `with:` statement let optimise out one of the contexts.

As suspicious as it is, the existing code already contains provisions against some optimisations, here:

https://github.com/Ericsson/codechecker/pull/4089/files#diff-857761db3a331e1e83bd931edaf83b01725c9b2c8d755a3d6257334ffad17d4b

I was not able to reproduce, neither pro nor contra, what this instruction was trying to prevent, and the blame history ran dry quickly as well, so I opted to keep the existing behaviour intact, even if hidden behind a small refactoring.

Reply:
Sorry, I don't really understand the situation. Unfortunately the link doesn't open for me properly, because GitHub collapses the files with too many changes. So could you, please, elaborate a little more? Do we need this noop if it doesn't reproduce the issue, then? :smile:

Reply:
@bruntib Here is the mentioned logic in the release version: https://github.com/Ericsson/codechecker/blob/v6.23.0/web/server/codechecker_server/api/mass_store_run.py#L1259-L1276

I would prefer to keep this noop, because I can neither prove nor disprove that removing it would result in the issue not reappearing. It seems that this behaviour is tied to some specifics of concurrent execution over multiple servers, something that is not simple enough to reproduce locally.

It seems it was I who introduced this in 17ddab02615fec2dd24733a2df9e76cb0a953f9f, but it was 7 years ago, and I do not remember contexts that far back in time.